autogenerate KMS CryptoKey, add purpose

drebes opened the pull request:
<!-- 
Note: You may see "This branch is out-of-date with the base branch"
when you submit a pull request. This is fine! We don't use the GitHub
merge button to merge PRs, and you can safely ignore that message.
-->

<!-- Your regular pull request description goes here -->




<!--
For each repository you expect to modify with this PR, fill in a repo-specific
PR title under the corresponding tag. We use repo-specified PR titles to ensure
that each downstream has a clear, easy to understand history.

If the Magician generates a PR for a repo with no specified title, it will use
the title of this PR. [terraform-beta] will inherit the title of [terraform]
if it has no specified title.
-->

<!-- Optional PR titles that describe your downstream changes -->
-----------------------------------------------------------------
# [all]
autogenerate KMS CryptoKey, add purpose
## [terraform]
### [terraform-beta]
## [ansible]
## [inspec]

rileykarson commented:
Generation is broken due to a merge conflict @ master. That's being resolved by https://github.com/GoogleCloudPlatform/magic-modules/pull/1854, I'll kickstart this when that's done and get a reviewer assigned to this.

modular-magician commented:
Hi!  I'm the modular magician, I work on Magic Modules.
This PR seems not to have generated downstream PRs before, as of c76d97080013feacfcda6549d10690df54a09882. 
## Pull request statuses
No diff detected in Ansible.
No diff detected in Inspec.
## New Pull Requests
I built this PR into one or more new PRs on other repositories, and when those are closed, this PR will also be merged and closed.
depends: https://github.com/terraform-providers/terraform-provider-google-beta/pull/787 
depends: https://github.com/GoogleCloudPlatform/terraform-google-conversion/pull/93 
depends: https://github.com/terraform-providers/terraform-provider-google/pull/3749

drebes commented:
Just realized that this is missing support for nextRotationTime which is required when rotationPeriod is set. I'll work on this, no need to review it just yet.

modular-magician commented:
Hi!  I'm the modular magician, I work on Magic Modules.
I see that this PR has already had some downstream PRs generated. Any open downstreams are already updated to your most recent commit, ae96e0ab4dfbfe35537e6085df7b214b2ebb48f1. 
## Pull request statuses
terraform-provider-google-beta already has an open PR.
terraform-google-conversion already has an open PR.
terraform-provider-google already has an open PR.
No diff detected in Ansible.
No diff detected in Inspec.
## New Pull Requests
I didn't open any new pull requests because of this PR.

modular-magician commented:
Hi!  I'm the modular magician, I work on Magic Modules.
I see that this PR has already had some downstream PRs generated. Any open downstreams are already updated to your most recent commit, 59425ed94bf80353a0d258effff2c120a247328b. 
## Pull request statuses
terraform-provider-google-beta already has an open PR.
terraform-google-conversion already has an open PR.
terraform-provider-google already has an open PR.
No diff detected in Ansible.
No diff detected in Inspec.
## New Pull Requests
I didn't open any new pull requests because of this PR.

modular-magician commented:
Hi!  I'm the modular magician, I work on Magic Modules.
I see that this PR has already had some downstream PRs generated. Any open downstreams are already updated to your most recent commit, 8889e6dff9f9fbc828de67ce14588a2cc4e746b7. 
## Pull request statuses
terraform-provider-google-beta already has an open PR.
terraform-google-conversion already has an open PR.
terraform-provider-google already has an open PR.
No diff detected in Ansible.
No diff detected in Inspec.
## New Pull Requests
I didn't open any new pull requests because of this PR.

modular-magician commented:
Hi!  I'm the modular magician, I work on Magic Modules.
I see that this PR has already had some downstream PRs generated. Any open downstreams are already updated to your most recent commit, 940b26db53597ba7c19036863a69a35aea289928. 
## Pull request statuses
terraform-provider-google-beta already has an open PR.
terraform-google-conversion already has an open PR.
terraform-provider-google already has an open PR.
No diff detected in Ansible.
No diff detected in Inspec.
## New Pull Requests
I didn't open any new pull requests because of this PR.

modular-magician commented:
Hi!  I'm the modular magician, I work on Magic Modules.
I see that this PR has already had some downstream PRs generated. Any open downstreams are already updated to your most recent commit, b65cd7be7d9bc0e981c6781d553d13872b639737. 
## Pull request statuses
terraform-provider-google-beta already has an open PR.
terraform-google-conversion already has an open PR.
terraform-provider-google already has an open PR.
No diff detected in Ansible.
No diff detected in Inspec.
## New Pull Requests
I didn't open any new pull requests because of this PR.

modular-magician commented:
Hi!  I'm the modular magician, I work on Magic Modules.
I see that this PR has already had some downstream PRs generated. Any open downstreams are already updated to your most recent commit, 8f375d73876159db76f188bdbce978a8cb992a2d. 
## Pull request statuses
terraform-provider-google-beta already has an open PR.
terraform-google-conversion already has an open PR.
terraform-provider-google already has an open PR.
No diff detected in Ansible.
No diff detected in Inspec.
## New Pull Requests
I didn't open any new pull requests because of this PR.

danawillow commented on products/kms/terraform.yaml:
```diff
@@ -45,7 +45,54 @@ overrides: !ruby/object:Overrides::ResourceOverrides
+      versionTemplate: !ruby/object:Overrides::Terraform::PropertyOverride
+        default_from_api: true
+      versionTemplate.protectionLevel: !ruby/object:Overrides::Terraform::PropertyOverride
+        default_from_api: true
```
Why the change to this instead of a default?

drebes replied:
This does not need to be specified at creation time and the API will fill it with a default of `SOFTWARE`. Isn't that what `default_from_api` is for?

drebes replied:
Also, the defaults only apply to symmetric keys. Leaving it up to the API to fill the default value (for symmetric keys) make the error message more clear when you try to create an asymmetric key and do not specify a versionTemplate:
```
resource "google_kms_crypto_key" "my-key-4" {
  name            = "my-key-4"
  key_ring        = google_kms_key_ring.my_key_ring.self_link
  purpose  = "ASYMMETRIC_SIGN"

  lifecycle {
    prevent_destroy = true
  }
}
```
```
Terraform will perform the following actions:

  # google_kms_crypto_key.my-key-4 will be created
  + resource "google_kms_crypto_key" "my-key-4" {
      + id        = (known after apply)
      + key_ring  = "projects/drebes-playground-tf-test-f74b/locations/us-central1/keyRings/my-key-ring"
      + name      = "my-key-4"
      + purpose   = "ASYMMETRIC_SIGN"
      + self_link = (known after apply)

      + version_template {
          + algorithm        = (known after apply)
          + protection_level = (known after apply)
        }
    }

Plan: 1 to add, 0 to change, 0 to destroy.
```
then
```
google_kms_crypto_key.my-key-4: Creating...

Error: Error creating CryptoKey: googleapi: Error 400: Request field version_template.algorithm is required if purpose is not ENCRYPT_DECRYPT.

  on kms.tf line 49, in resource "google_kms_crypto_key" "my-key-4":
  49: resource "google_kms_crypto_key" "my-key-4" {
```
Otherwise you'd get an error saying that `GOOGLE_SYMMETRIC_ENCRYPTION`(which the user didn't explicitly specify) is not a valid algorithm for asymmetric keys. I find the error above more user friendly.

danawillow replied:
Yes, but that's also what defaults are for.

There isn't any real guidance on when to use one vs the other. Defaults have the advantage of showing up in the plan (vs <computed>), and fields with defaults can also be changed by removing the field from configs (vs computed fields, which don't show a diff if it isn't defined). Computed fields have the advantage of reacting appropriately if the API changes what the default value is.

My opinion generally for this sort of thing is that we shouldn't change things that aren't broken, but I'm also not going to block approval on it.

drebes replied:
Thanks for clarifying. Makes sense to change the protectionLevel to "SOFTWARE" since that always, apply, but I'll leave the algorithm as default_from_api since the old hardcoded default no longer applies when we're dealing with asymmetric keys. Pushed the changed.

danawillow replied:
Oh by the way, my comment got in a bit of a race condition with yours, so it should be read as having responded to your previous comment.

I'm a bit confused about your remarks regarding `algorithm`. Isn't that field marked as required?

drebes replied:
`algorithm` is required only if `versionTemplate` is specified, but `versionTemplate` is not required. See below an example:
```
POST /v1/projects/drebes-playground-tf-test-f74b/locations/us-central1/keyRings/my-key-ring/cryptoKeys?alt=json&cryptoKeyId=my-key-5&prettyPrint=false HTTP/1.1
Host: cloudkms.googleapis.com
User-Agent: google-api-go-client/0.5 Terraform/0.12.0 (+https://www.terraform.io) terraform-provider-google/dev
Content-Length: 30
Content-Type: application/json
Accept-Encoding: gzip

{
 "purpose": "ENCRYPT_DECRYPT"
}

-----------------------------------------------------
2019/06/13 00:36:23 [DEBUG] Google API Response Details:
---[ RESPONSE ]--------------------------------------
HTTP/2.0 200 OK
Alt-Svc: quic=":443"; ma=2592000; v="46,44,43,39"
Cache-Control: private
Content-Type: application/json; charset=UTF-8
Date: Wed, 12 Jun 2019 22:36:23 GMT
Server: ESF
Vary: Origin
Vary: X-Origin
Vary: Referer
X-Content-Type-Options: nosniff
X-Frame-Options: SAMEORIGIN
X-Xss-Protection: 0

{
 "name": "projects/drebes-playground-tf-test-f74b/locations/us-central1/keyRings/my-key-ring/cryptoKeys/my-key-5",
 "primary": {
  "name": "projects/drebes-playground-tf-test-f74b/locations/us-central1/keyRings/my-key-ring/cryptoKeys/my-key-5/cryptoKeyVersions/1",
  "state": "ENABLED",
  "createTime": "2019-06-12T22:36:23.559829707Z",
  "protectionLevel": "SOFTWARE",
  "algorithm": "GOOGLE_SYMMETRIC_ENCRYPTION",
  "generateTime": "2019-06-12T22:36:23.559829707Z"
 },
 "purpose": "ENCRYPT_DECRYPT",
 "createTime": "2019-06-12T22:36:23.559829707Z",
 "versionTemplate": {
  "protectionLevel": "SOFTWARE",
  "algorithm": "GOOGLE_SYMMETRIC_ENCRYPTION"
 }
}
-----------------------------------------------------
```

danawillow commented on products/kms/api.yaml:
```diff
@@ -56,3 +56,74 @@ objects:
+               The algorithm to use when creating a version based on this template.
+               See the [algorithm reference](https://cloud.google.com/kms/docs/reference/rest/v1/CryptoKeyVersionAlgorithm) for possible inputs.
+            required: true
+            input: true
```
This was previously not ForceNew- was that a bug, or is this an oversight?

drebes replied:
It was effectively a ForceNew before since `version_template` has only `algorithm` and `protection_level`,  `protection_level` is never updatable (so ForceNew), and `algorithm` can only be `GOOGLE_SYMMETRIC_ENCRYPTION` for symmetric keys. But with asymmetric keys which we will support now with `purpose` we should be able to update `algorithm`, you're right. I'll fix this.

drebes replied:
Interesting, got hit by the old `updateMask` limitations which my custom `update_mask_fields` fortunately also works around:
```
  # google_kms_crypto_key.my-key-2 will be updated in-place
  ~ resource "google_kms_crypto_key" "my-key-2" {
        id        = "projects/drebes-playground-tf-test-f74b/locations/us-central1/keyRings/my-key-ring/cryptoKeys/crypto-key-2"
        key_ring  = "projects/drebes-playground-tf-test-f74b/locations/us-central1/keyRings/my-key-ring"
        name      = "crypto-key-2"
        purpose   = "ASYMMETRIC_DECRYPT"
        self_link = "projects/drebes-playground-tf-test-f74b/locations/us-central1/keyRings/my-key-ring/cryptoKeys/crypto-key-2"

        timeouts {}

      ~ version_template {
          ~ algorithm        = "RSA_DECRYPT_OAEP_3072_SHA256" -> "RSA_DECRYPT_OAEP_4096_SHA256"
            protection_level = "HSM"
        }
    }

Plan: 0 to add, 1 to change, 0 to destroy.
```
standard PATCH fails:
```
PATCH /v1/projects/drebes-playground-tf-test-f74b/locations/us-central1/keyRings/my-key-ring/cryptoKeys/crypto-key-2?alt=json&updateMask=versionTemplate HTTP/1.1
Host: cloudkms.googleapis.com
User-Agent: Terraform/0.12.0 (+https://www.terraform.io) terraform-provider-google/dev
Content-Length: 89
Content-Type: application/json
Accept-Encoding: gzip

{
 "versionTemplate": {
  "algorithm": "RSA_DECRYPT_OAEP_4096_SHA256",
  "protectionLevel": "HSM"
 }
}

-----------------------------------------------------
2019/06/12 23:17:53 [DEBUG] Google API Response Details:
---[ RESPONSE ]--------------------------------------
HTTP/2.0 400 Bad Request
Alt-Svc: quic=":443"; ma=2592000; v="46,44,43,39"
Cache-Control: private
Content-Type: application/json; charset=UTF-8
Date: Wed, 12 Jun 2019 21:17:53 GMT
Server: ESF
Vary: Origin
Vary: X-Origin
Vary: Referer
X-Content-Type-Options: nosniff
X-Frame-Options: SAMEORIGIN
X-Xss-Protection: 0

{
  "error": {
    "code": 400,
    "message": "Field crypto_key.version_template.protection_level may not be updated. The value is immutable.",
    "status": "INVALID_ARGUMENT"
  }
}

-----------------------------------------------------
```
but succeeds with a custom `update_mask_fields`:
```
PATCH /v1/projects/drebes-playground-tf-test-f74b/locations/us-central1/keyRings/my-key-ring/cryptoKeys/crypto-key-2?alt=json&updateMask=versionTemplate.algorithm HTTP/1.1
Host: cloudkms.googleapis.com
User-Agent: Terraform/0.12.0 (+https://www.terraform.io) terraform-provider-google/dev
Content-Length: 89
Content-Type: application/json
Accept-Encoding: gzip

{
 "versionTemplate": {
  "algorithm": "RSA_DECRYPT_OAEP_4096_SHA256",
  "protectionLevel": "HSM"
 }
}

-----------------------------------------------------
2019/06/12 23:35:57 [DEBUG] Google API Response Details:
---[ RESPONSE ]--------------------------------------
HTTP/2.0 200 OK
Alt-Svc: quic=":443"; ma=2592000; v="46,44,43,39"
Cache-Control: private
Content-Type: application/json; charset=UTF-8
Date: Wed, 12 Jun 2019 21:35:57 GMT
Server: ESF
Vary: Origin
Vary: X-Origin
Vary: Referer
X-Content-Type-Options: nosniff
X-Frame-Options: SAMEORIGIN
X-Xss-Protection: 0

{
  "name": "projects/drebes-playground-tf-test-f74b/locations/us-central1/keyRings/my-key-ring/cryptoKeys/crypto-key-2",
  "purpose": "ASYMMETRIC_DECRYPT",
  "createTime": "2019-06-02T14:15:09.913833168Z",
  "versionTemplate": {
    "protectionLevel": "HSM",
    "algorithm": "RSA_DECRYPT_OAEP_4096_SHA256"
  }
}

-----------------------------------------------------
```
Pushed my changes now.

danawillow commented on templates/terraform/custom_import/kms_crypto_key.go.erb:
```diff
@@ -0,0 +1,24 @@
+
+	config := meta.(*Config)
+
+	// import ID is in the form `{{keyRing}}/cryptoKeys/{{name}}` and `keyRing` is in the
```
Could you use the existing parseKmsCryptoKeyId fn here instead? That has the benefit of not requiring the project

drebes replied:
Will look into it. `{{project}}/{{keyring_name}}/{{key_name}}` indeed seems much simpler to specify.

drebes replied:
Done.

danawillow commented on templates/terraform/decoders/kms_crypto_key.go.erb:
```diff
@@ -0,0 +1,17 @@
+	# limitations under the License.
+-%>
+	// Take the returned long form of the name and use it as `self_link`.
+	d.Set("self_link", res["name"].(string))
```
Is it possible to reuse the decoder from keyring here?

drebes replied:
Seems harmless, `res["name"] = d.Get("name").(string)` will have no effect since we don't do anything with the returned name for CryptoKey, though. I'll reuse it.

drebes replied:
Done.

modular-magician commented:
Hi!  I'm the modular magician, I work on Magic Modules.
I see that this PR has already had some downstream PRs generated. Any open downstreams are already updated to your most recent commit, 4c6a4120ecf4120e4022fea147ea5ea1b07fd342. 
## Pull request statuses
terraform-provider-google-beta already has an open PR.
terraform-google-conversion already has an open PR.
terraform-provider-google already has an open PR.
No diff detected in Ansible.
No diff detected in Inspec.
## New Pull Requests
I didn't open any new pull requests because of this PR.

modular-magician commented:
Hi!  I'm the modular magician, I work on Magic Modules.
I see that this PR has already had some downstream PRs generated. Any open downstreams are already updated to your most recent commit, 0cc2274b8c4e3148791aa8b83a31a6896135c6d4. 
## Pull request statuses
terraform-provider-google-beta already has an open PR.
terraform-google-conversion already has an open PR.
terraform-provider-google already has an open PR.
No diff detected in Ansible.
No diff detected in Inspec.
## New Pull Requests
I didn't open any new pull requests because of this PR.

danawillow commented on templates/terraform/update_mask.erb:
```diff
@@ -1,5 +1,4 @@
 updateMask := []string{}
-
 <% update_body_properties.each do |prop| -%>
 <%# UpdateMask documentation is not not obvious about which fields are supported or
```
Looks like this comment is out of date now

drebes replied:
Updated.

modular-magician commented:
Hi!  I'm the modular magician, I work on Magic Modules.
I see that this PR has already had some downstream PRs generated. Any open downstreams are already updated to your most recent commit, d5ac81795f75669f3701a29071ebb60734aeba38. 
## Pull request statuses
terraform-provider-google-beta already has an open PR.
terraform-google-conversion already has an open PR.
terraform-provider-google already has an open PR.
No diff detected in Ansible.
No diff detected in Inspec.
## New Pull Requests
I didn't open any new pull requests because of this PR.

modular-magician commented:
Hi!  I'm the modular magician, I work on Magic Modules.
I see that this PR has already had some downstream PRs generated. Any open downstreams are already updated to your most recent commit, 1680559c4b2055f2f0bd57debe2d80a27823c26b. 
## Pull request statuses
terraform-provider-google-beta already has an open PR.
terraform-google-conversion already has an open PR.
terraform-provider-google already has an open PR.
No diff detected in Ansible.
No diff detected in Inspec.
## New Pull Requests
I didn't open any new pull requests because of this PR.

modular-magician commented:
Hi!  I'm the modular magician, I work on Magic Modules.
I see that this PR has already had some downstream PRs generated. Any open downstreams are already updated to your most recent commit, d9b62c18b69dcb0a628b2edd6e0dcbcc11ff86e2. 
## Pull request statuses
terraform-provider-google-beta already has an open PR.
terraform-google-conversion already has an open PR.
terraform-provider-google already has an open PR.
No diff detected in Ansible.
No diff detected in Inspec.
## New Pull Requests
I didn't open any new pull requests because of this PR.

modular-magician commented:
Hi!  I'm the modular magician, I work on Magic Modules.
I see that this PR has already had some downstream PRs generated. Any open downstreams are already updated to your most recent commit, 7953bf3cb0d75dbdfe10cc3810a19c5e138d1c12. 
## Pull request statuses
terraform-provider-google-beta already has an open PR.
terraform-google-conversion already has an open PR.
terraform-provider-google already has an open PR.
No diff detected in Ansible.
No diff detected in Inspec.
## New Pull Requests
I didn't open any new pull requests because of this PR.

drebes commented:
I've managed to set my test environment to create new projects so I've also added some new acceptance tests:
```
$ make testacc TEST=./google TESTARGS='-run=TestAccKmsCryptoKey' 
==> Checking source code against gofmt...
==> Checking that code complies with gofmt requirements...
TF_ACC=1 TF_SCHEMA_PANIC_ON_ERROR=1 go test ./google -v -run=TestAccKmsCryptoKey -timeout 240m -ldflags="-X=github.com/terraform-providers/terraform-provider-google/version.ProviderVersion=acc"
=== RUN   TestAccKmsCryptoKeyIamBinding
=== PAUSE TestAccKmsCryptoKeyIamBinding
=== RUN   TestAccKmsCryptoKeyIamMember
=== PAUSE TestAccKmsCryptoKeyIamMember
=== RUN   TestAccKmsCryptoKey_basic
=== PAUSE TestAccKmsCryptoKey_basic
=== RUN   TestAccKmsCryptoKey_rotation
=== PAUSE TestAccKmsCryptoKey_rotation
=== RUN   TestAccKmsCryptoKey_template
=== PAUSE TestAccKmsCryptoKey_template
=== CONT  TestAccKmsCryptoKeyIamBinding
=== CONT  TestAccKmsCryptoKey_rotation
=== CONT  TestAccKmsCryptoKey_template
=== CONT  TestAccKmsCryptoKey_basic
--- PASS: TestAccKmsCryptoKey_basic (54.50s)
=== CONT  TestAccKmsCryptoKeyIamMember
--- PASS: TestAccKmsCryptoKey_template (56.28s)
--- PASS: TestAccKmsCryptoKey_rotation (65.10s)
--- PASS: TestAccKmsCryptoKeyIamBinding (82.73s)
--- PASS: TestAccKmsCryptoKeyIamMember (70.81s)
PASS
ok  	github.com/terraform-providers/terraform-provider-google/google	125.358s
```
I also triggered a problem with the existing tests. There was no test with no rotationPeriod set, and we were not checking if rotationPeriod was already disabled when we try to disable it at key delete, which caused an error. I fixed that as well.

modular-magician commented:
Hi!  I'm the modular magician, I work on Magic Modules.
I see that this PR has already had some downstream PRs generated. Any open downstreams are already updated to your most recent commit, 841d84ac6ef2dc360d74def24af5b8905233354f. 
## Pull request statuses
terraform-provider-google-beta already has an open PR.
terraform-google-conversion already has an open PR.
terraform-provider-google already has an open PR.
No diff detected in Ansible.
No diff detected in Inspec.
## New Pull Requests
I didn't open any new pull requests because of this PR.
